src/voice/udp/handle.rs

```diff
@@ -0,0 +1,256 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::sync::Arc;
+
+use crypto_secretbox::{
+    aead::Aead, cipher::generic_array::GenericArray, KeyInit, XSalsa20Poly1305,
+};
+use discortp::Packet;
+
+use getrandom::getrandom;
+use log::*;
+
+use tokio::{sync::Mutex, sync::RwLock};
+
+use super::UdpSocket;
+
+use crate::{
+    errors::VoiceUdpError,
+    types::VoiceEncryptionMode,
+    voice::{crypto::get_xsalsa20_poly1305_nonce, voice_data::VoiceData},
+};
+
+use super::{events::VoiceUDPEvents, RTP_HEADER_SIZE};
+
+/// Handle to a voice udp connection
+///
+/// Can be safely cloned and will still correspond to the same connection.
+#[derive(Debug, Clone)]
+pub struct UdpHandle {
+    pub events: Arc<Mutex<VoiceUDPEvents>>,
+    pub(super) socket: Arc<UdpSocket>,
+    pub data: Arc<RwLock<VoiceData>>,
+}
+
+impl UdpHandle {
+    /// Constructs and sends encoded opus rtp data.
+    ///
+    /// Automatically makes an [RtpPacket](discortp::rtp::RtpPacket), encrypts it and sends it.
+    ///
+    /// # Errors
+    /// If we do not have VoiceReady data, which contains our ssrc, this returns a
+    /// [VoiceUdpError::NoData] error.
+    ///
+    /// If we have not received an encryption key, this returns a [VoiceUdpError::NoKey] error.
+    ///
+    /// If the Udp socket is broken, this returns a [VoiceUdpError::BrokenSocket] error.
```

Comment:
<!--
 Copyright (c) 2024 star
 
 This Source Code Form is subject to the terms of the Mozilla Public
 License, v. 2.0. If a copy of the MPL was not distributed with this
 file, You can obtain one at https://mozilla.org/MPL/2.0/.
-->


```suggestion
    /// If the UDP socket is broken, this returns a [VoiceUdpError::BrokenSocket] error.
```